dev/Security.AccessControl/Security.AccessControl/GetSecurityDescriptorForAppContainerNames.cpp

```diff
@@ -0,0 +1,116 @@
+﻿#include "pch.h"
+
+#include <wil/resource.h>
+#include <wil/result.h>
+#include "Security.AccessControl.h"
+#include <vector>
+#include <sddl.h>
+#include <AclAPI.h>
+#include <UserEnv.h>
+
+using namespace std;
+
+namespace
+{
+    wil::unique_any_psid GetCurrentProcessSid()
+    {
+        wil::unique_any_psid result;
+        wil::unique_process_heap_ptr<TOKEN_GROUPS> tg;
+        DWORD length = 0;
+        if (!::GetTokenInformation(
+            ::GetCurrentProcessToken(),
+            TokenLogonSid,
+            nullptr,
+            0,
+            &length
+        ))
+        {
+            auto err = ::GetLastError();
+            if (err != ERROR_INSUFFICIENT_BUFFER)
+            {
+                THROW_WIN32(err);
+            }
+            tg.reset(static_cast<TOKEN_GROUPS*>(::HeapAlloc(::GetProcessHeap(), HEAP_ZERO_MEMORY, length)));
+        }
+
+        THROW_IF_WIN32_BOOL_FALSE(::GetTokenInformation(
+            ::GetCurrentProcessToken(),
+            TokenLogonSid,
+            tg.get(),
+            length,
+            &length));
+
+        auto sidLength = ::GetLengthSid(tg->Groups[0].Sid);
+        result.reset(::HeapAlloc(::GetProcessHeap(), 0, sidLength));
+        THROW_IF_WIN32_BOOL_FALSE(::CopySid(
+            sidLength, result.get(), tg->Groups[0].Sid));
+        return result;
+    }
+}
+
+STDAPI GetSecurityDescriptorForAppContainerNames(
+    uint32_t accessRequestCount,
+    _In_reads_(accessRequestCount)
+        const AppContainerNameAndAccess* accessRequests,
+    _In_opt_ PSID principal,
+    uint32_t principalAccessMask,
+    _Outptr_ PSECURITY_DESCRIPTOR* securityDescriptor
+)
```

Comment:
Use a function-level try/catch to reduce one level of indentation.  This should probably be marked noexcept as well.

Reply:
This function is the implementation of a "Flat C" API. So we have a larger question of whether we are going to mark those APIs as `noexcept` in the headers.

Clearly, that keyword won't work for plain C consumers. It's only available for C++ consumers if they are running C++11 or greater. I'm in favor of this, but not sure if this decision has been made yet.

Reply:
I have moved the API into `extern "C"` which has a similar benefit to `noexcept` when compiling with `/EHsc` (which is the recommended setting).